packages/aws-cdk-lib/aws-stepfunctions/lib/state-machine.ts

```diff
@@ -452,6 +476,39 @@ export class StateMachine extends StateMachineBase {
       }
     }
 
+    if (props?.kmsKey) {
+      this.role.addToPrincipalPolicy(new iam.PolicyStatement({
+        effect: iam.Effect.ALLOW,
+        actions: [
+          'kms:Decrypt', 'kms:GenerateDataKey',
+        ],
+        resources: [`${props.kmsKey.keyArn}`],
+      }));
+
+      props.kmsKey.addToResourcePolicy(new iam.PolicyStatement({
+        resources: ['*'],
```

Comment:
Since this is a resource based policy, the `*` refers to the key itself when attaching this policy: https://docs.aws.amazon.com/kms/latest/developerguide/iam-policies-best-practices.html

> In a key policy, the wildcard character in the Resource element represents the KMS key to which the key policy is attached